src/anyp/Uri.cc

```diff
@@ -461,24 +444,60 @@ AnyP::Uri::parse(const HttpRequestMethod& method, const char *url)
         }
     }
 
-    parseFinish(protocol, proto, urlpath, foundHost, SBuf(login), foundPort);
+    setScheme(scheme);
+    path(urlpath);
+    host(foundHost);
+    userInfo(SBuf(login));
+    port(foundPort);
     return true;
+
+    } catch (...) {
+        debugs(23, 2, "error: " << CurrentException << Raw("rawUrl", rawUrl.rawContent(), rawUrl.length()).minLevel(DBG_DATA));
+        return false;
+    }
 }
 
-/// Update the URL object with parsed URI data.
+/**
+ * Governed by RFC 8141 section 2:
+ *
+ *  assigned-name = "urn" ":" NID ":" NSS
+ *  NID           = (alphanum) 0*30(ldh) (alphanum)
+ *  ldh           = alphanum / "-"
+ *  NSS           = pchar *(pchar / "/")
+ *
+ * RFC 3986 Appendix D.2 defines (as deprecated):
+ *
+ *   alphanum     = ALPHA / DIGIT
+ *
+ * Notice that NID is exactly 2-32 characters in length.
+ */
 void
-AnyP::Uri::parseFinish(const AnyP::ProtocolType protocol,
-                       const char *const protoStr, // for unknown protocols
-                       const char *const aUrlPath,
-                       const char *const aHost,
-                       const SBuf &aLogin,
-                       const int aPort)
+AnyP::Uri::parseUrn(Parser::Tokenizer &tok)
 {
-    setScheme(protocol, protoStr);
-    path(aUrlPath);
-    host(aHost);
-    userInfo(aLogin);
-    port(aPort);
+    static const auto nidChars = CharacterSet("NID","-") + CharacterSet::ALPHA + CharacterSet::DIGIT;
+    static const auto alphanum = (CharacterSet::ALPHA + CharacterSet::DIGIT).rename("alphanum");
+    SBuf nid;
+    if (!tok.prefix(nid, nidChars, 32))
+        throw TextException("NID not found", Here());
+
+    if (!tok.skip(':'))
+        throw TextException("NID too long or missing ':' delimiter", Here());
+
+    debugs(23, 3, "Split URI into proto='urn', nid='" << nid << "', path='" << tok.remaining() << "'");
```

Comment:
AFAICT, `tok.remaining()` can be huge (now or in the future optimized code) -- we may not be dealing with just the URL input here but a whole POST request, for example.

* If you really want to dump remaining input here, then please use the Raw class. The Raw class already have a 40-character limit for non-DATA dumping which would work well here.
* However, it may be best to leave input printing to your exception catcher instead -- it will cover a lot more problematic cases than this particular debug statement can. And the catcher can print the prefix of the failed input in addition to the often less important suffix.

Reply:
Done the Raw use. This is printing the parse action of what the URI got split into, not errors. So leaving it in this method. Though I have now moved it to only happen on the success case which the exception handler output does not cover.

Reply:
> This is printing the parse action of what the URI got split into, not errors. So leaving it in this method

Please note that parseUrn() description does not specify whether the method needs extracted/isolated URI or any input that _starts_ with a URI. Since a lot of parseFoo() functions extract their Foos from input while leaving the remaining input intact, consider adjusting parseUrn() description to mention that isolation requirement.

I do not insist on this change and am resolving this change request.